File: Week 5/공통문제 5.py
'''
[
[
"POOOP",
"OXXOX",
"OPXPX",
"OOXOX",
"POXXP",]
1. P일때만 검사
2. 사방향 P
3. 한방향 이상 O  > 그 O에서 +1 -1씩 해서 P가 존재하면 false
4. X면 그냥 넘기기

'''
def solution(places):
    
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    answer = []   
    def check(test):
        for i in range(5):
            for j in range(5):
                if test[i][j] == 'P':
                    for k in range(4):
                        a = i + dx[k]
                        b = j + dy[k]
                        if a < 0 or a >= 5 or b < 0 or b >= 5:
                            continue
                        if test[a][b] == 'P':
                            return False
                        if test[a][b] == 'O':
                            count = 0
                            for l in range(4):
                                c = a + dx[l]
                                d = b + dy[l]
                                if c < 0 or c >= 5 or d < 0 or d >= 5:
                                    continue
                                if test[c][d] == 'P':
                                    count += 1
                                    if count >= 2:
                                        count = 0
                                        return False
        return True
    
    for test in places:
        isTrue = check(test)
        if(isTrue):
            answer.append(1)
        else:
            answer.append(0)
                                        
        
           
    
    return answer

File: Week 5/공통문제 5.js
// 현재 위치가 P일때만 검사하기
// 상하좌우가 O인경우에 그쪽 인덱스로 이동해서 상하좌우 P count가 2 이상이면 false
// X인경우는 넘어가기
// P일때 상하좌우가 P인경우도 false

function solution(places) {
    var answer = [];

    function checkGrp(grp){
        let dx = [1,-1,0,0]
        let dy = [0,0,1,-1]
        for (var i = 0 ; i < 5 ; i++){
            for (var j = 0 ; j < 5 ; j++){
                if(grp[i][j] == "P"){

                    for (var k = 0 ; k < 4 ; k++){
                        let a = i + dx[k]
                        let b = j + dy[k]
                        if (a < 0 || a >= 5 || b < 0 || b >= 5){
                            continue
                        }
                        if (grp[a][b] == "O"){
                            //상하좌우에 O가 있다
                            var pCount = 0
                            for (var l = 0 ; l < 4 ; l++){
                                let n = a + dx[l]
                                let m = b + dy[l]
                                if (n < 0 || n >= 5 || m < 0 || m >= 5){
                                    continue
                                }
                                if( grp[n][m] == "P" ){
                                    pCount += 1
                                    if(pCount >= 2){
                                        return false
                                    }
                                }
                            }
                        }
                        else if (grp[a][b] == "P"){
                            return false
                        }
                    }
                }
            }
        }
        return true
    }
    places.forEach(ele => {
        if(checkGrp(ele)){
            answer.push(1)
        }else{
            answer.push(0)
        }
    })
    return answer;
}
